ide upgrade fails

### Actual behavior

```
$ ide upgrade --mode=stable
Trying to download ide-cli-2025.08.001-windows-x64.tar.gz from https://repo1.maven.org/maven2/com/devonfw/tools/IDEasy/ide-cli/2025.08.001/ide-cli-2025.08.001-windows-x64.tar.gz
Downloading 100% [=============================================================] 22/22MiB (0:00:01 / 0:00:00) 22.9MiB/s
Trying to download ide-cli-2025.08.001-windows-x64.tar.gz.md5 from https://repo1.maven.org/maven2/com/devonfw/tools/IDEasy/ide-cli/2025.08.001/ide-cli-2025.08.001-windows-x64.tar.gz.md5
Trying to download ide-cli-2025.08.001-windows-x64.tar.gz.sha1 from https://repo1.maven.org/maven2/com/devonfw/tools/IDEasy/ide-cli/2025.08.001/ide-cli-2025.08.001-windows-x64.tar.gz.sha1
MD5 checksum b5962c3fd15c755021d7a447ef7ae011 is correct.
SHA1 checksum 71e016fa64ba04fa7a8aec1600fac9d5eb6630b1 is correct.
Your IDEasy installation is missing the version file.
Extracting TAR file C:\Users\joerg\.m2\repository\com\devonfw\tools\IDEasy\ide-cli\2025.08.001\ide-cli-2025.08.001-windows-x64.tar.gz to C:\Users\joerg\projects\_ide\tmp\extract-ide-cli-2025.08.001-windows-x64.tar.gz
Extracting 100% [=================================================================] 22/22MiB (0:00:00 / 0:00:00) ?MiB/s
An unexpected error occurred!
We are sorry for the inconvenience.
Please check the error below, resolve it and try again.
If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:
https://github.com/devonfw/IDEasy/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=IllegalStateException%3A+Failed+to+move+directory%3A+C%3A%5CUsers%5Cjoerg%5Cprojects%5C_ide%5Ctmp%5Cextract-ide-cli-2025.08.001-windows-x64.tar.gz+to+C%3A%5CUsers%5Cjoerg%5Cprojects%5C_ide%5Csoftware%5Cmaven%5Cideasy%5Cideasy%5C2025.08.001.%0AOn+Windows%2C+file+operations+could+fail+due+to+file+locks.+Please+ensure+the+files+in+the+moved+directory+are+not+in+use.+For+further+details%2C+see%3A+%0Ahttps%3A%2F%2Fgithub.com%2Fdevonfw%2FIDEasy%2Fblob%2Fmain%2Fdocumentation%2Fwindows-file-lock.adoc
java.lang.IllegalStateException: Failed to move directory: C:\Users\joerg\projects\_ide\tmp\extract-ide-cli-2025.08.001-windows-x64.tar.gz to C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.08.001.
On Windows, file operations could fail due to file locks. Please ensure the files in the moved directory are not in use. For further details, see:
https://github.com/devonfw/IDEasy/blob/main/documentation/windows-file-lock.adoc
        at com.devonfw.tools.ide.io.FileAccessImpl.move(FileAccessImpl.java:364)
        at com.devonfw.tools.ide.io.FileAccessImpl.extract(FileAccessImpl.java:627)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.performToolInstallation(LocalToolCommandlet.java:247)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:213)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:178)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.doInstallStep(LocalToolCommandlet.java:91)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.install(LocalToolCommandlet.java:81)
        at com.devonfw.tools.ide.tool.ToolCommandlet.install(ToolCommandlet.java:264)
        at com.devonfw.tools.ide.tool.IdeasyCommandlet.install(IdeasyCommandlet.java:122)
        at com.devonfw.tools.ide.commandlet.UpgradeCommandlet.run(UpgradeCommandlet.java:43)
        at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:1110)
        at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:1020)
        at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:95)
        at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:59)
        at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:133)
        at java.base@23.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: java.nio.file.FileAlreadyExistsException: C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.08.001
        at java.base@23.0.2/sun.nio.fs.WindowsFileCopy.move(WindowsFileCopy.java:374)
        at java.base@23.0.2/sun.nio.fs.WindowsFileSystemProvider.move(WindowsFileSystemProvider.java:291)
        at java.base@23.0.2/java.nio.file.Files.move(Files.java:1437)
        at com.devonfw.tools.ide.io.FileAccessImpl.move(FileAccessImpl.java:357)
        ... 15 more


Error: IDEasy failed with exit code 255
```

### Reproduce

1. upgrade to the latest stable release (`ide upgrade`)
2. switch to latest snapshot (`ide upgrade --mode=snapshot`)
3. switch back (`ide upgrade --mode=stable`)
4. see error above

### Expected behavior

Switching back should work seamless.

### IDEasy status

```bash
IDE_ROOT is set to ~/projects
You are not inside an IDE project: ~/projects
You are online.
Your version of IDEasy is 2025.09.001-09_17_02-SNAPSHOT.
You are using a SNAPSHOT version of IDEasy. For stability consider switching to a stable release via 'ide upgrade --mode=stable'
Your version of IDEasy is 2025.09.001-09_17_02-SNAPSHOT but version 2025.09.001-20250917.023913-3 is available. Please run the following command to upgrade to the latest version:
ide upgrade
Your operating system is windows(10.0)@x64 [Windows 11@amd64]
Successfully completed ide (status)
```

### Related/Dependent issues

_No response_

### Comments/Hints

_No response_

--- a/cli/src/main/java/com/devonfw/tools/ide/property/Property.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/property/Property.java
@@ -242,6 +242,9 @@ public abstract class Property<V> {
    */
   public void setValue(V value) {
 
+    if (!this.multivalued) {
+      this.value.clear();
+    }
     this.value.add(value);
   }
 

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
@@ -63,6 +63,15 @@ public abstract class LocalToolCommandlet extends ToolCommandlet {
   }
 
   /**
+   * @return {@code true} to ignore a missing {@link IdeContext#FILE_SOFTWARE_VERSION software version file} in an installation, {@code false} delete the broken
+   *     installation (default).
+   */
+  protected boolean isIgnoreMissingSoftwareVersionFile() {
+
+    return false;
+  }
+
+  /**
    * @deprecated will be removed once all "dependencies.json" are created in ide-urls.
    */
   @Deprecated
@@ -215,7 +224,7 @@ public abstract class LocalToolCommandlet extends ToolCommandlet {
     if (Files.isDirectory(installationPath)) {
       if (Files.exists(toolVersionFile)) {
         if (!ignoreSoftwareRepo) {
-          assert installedVersion.toString().equals(fileAccess.readFileContent(toolVersionFile));
+          assert resolvedVersion.toString().equals(fileAccess.readFileContent(toolVersionFile));
           this.context.debug("Version {} of tool {} is already installed at {}", resolvedVersion, getToolWithEdition(this.tool, edition), installationPath);
           return createToolInstallation(installationPath, resolvedVersion, false, processContext, extraInstallation);
         }
@@ -224,8 +233,7 @@ public abstract class LocalToolCommandlet extends ToolCommandlet {
         if (this.tool.equals(IdeasyCommandlet.TOOL_NAME)) {
           this.context.warning("Your IDEasy installation is missing the version file at {}", toolVersionFile);
           return createToolInstallation(installationPath, resolvedVersion, false, processContext, extraInstallation);
-        } else {
-          // TODO: fixme for NPM based tools
+        } else if (!isIgnoreMissingSoftwareVersionFile()) {
           this.context.warning("Deleting corrupted installation at {}", installationPath);
           fileAccess.delete(installationPath);
         }

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
@@ -224,7 +224,8 @@ public abstract class LocalToolCommandlet extends ToolCommandlet {
     if (Files.isDirectory(installationPath)) {
       if (Files.exists(toolVersionFile)) {
         if (!ignoreSoftwareRepo) {
-          assert resolvedVersion.toString().equals(fileAccess.readFileContent(toolVersionFile));
+          assert resolvedVersion.equals(getInstalledVersion(installationPath)) :
+              "Found version " + getInstalledVersion(installationPath) + " in " + toolVersionFile + " but expected " + resolvedVersion;
           this.context.debug("Version {} of tool {} is already installed at {}", resolvedVersion, getToolWithEdition(this.tool, edition), installationPath);
           return createToolInstallation(installationPath, resolvedVersion, false, processContext, extraInstallation);
         }

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
@@ -210,6 +210,27 @@ public abstract class LocalToolCommandlet extends ToolCommandlet {
       this.context.debug("Version {} of tool {} is already installed at {}", resolvedVersion, getToolWithEdition(this.tool, edition), installationPath);
       return createToolInstallation(installationPath, resolvedVersion, false, processContext, extraInstallation);
     }
+    Path toolVersionFile = installationPath.resolve(IdeContext.FILE_SOFTWARE_VERSION);
+    FileAccess fileAccess = this.context.getFileAccess();
+    if (Files.isDirectory(installationPath)) {
+      if (Files.exists(toolVersionFile)) {
+        if (!ignoreSoftwareRepo) {
+          assert installedVersion.toString().equals(fileAccess.readFileContent(toolVersionFile));
+          this.context.debug("Version {} of tool {} is already installed at {}", resolvedVersion, getToolWithEdition(this.tool, edition), installationPath);
+          return createToolInstallation(installationPath, resolvedVersion, false, processContext, extraInstallation);
+        }
+      } else {
+        // Makes sure that IDEasy will not delete itself
+        if (this.tool.equals(IdeasyCommandlet.TOOL_NAME)) {
+          this.context.warning("Your IDEasy installation is missing the version file at {}", toolVersionFile);
+          return createToolInstallation(installationPath, resolvedVersion, false, processContext, extraInstallation);
+        } else {
+          // TODO: fixme for NPM based tools
+          this.context.warning("Deleting corrupted installation at {}", installationPath);
+          fileAccess.delete(installationPath);
+        }
+      }
+    }
     performToolInstallation(toolRepository, resolvedVersion, installationPath, edition, processContext);
     return createToolInstallation(installationPath, resolvedVersion, true, processContext, extraInstallation);
   }

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
@@ -4,7 +4,6 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
-import java.util.Objects;
 import java.util.Set;
 
 import com.devonfw.tools.ide.commandlet.Commandlet;
@@ -378,8 +377,6 @@ public abstract class ToolCommandlet extends Commandlet implements Tags {
 
     String edition = getConfiguredEdition();
     ToolRepository toolRepository = getToolRepository();
-    VersionIdentifier versionIdentifier = toolRepository.resolveVersion(this.tool, edition, version, this);
-    Objects.requireNonNull(versionIdentifier);
 
     EnvironmentVariables variables = this.context.getVariables();
     if (destination == null) {
@@ -389,13 +386,9 @@ public abstract class ToolCommandlet extends Commandlet implements Tags {
     EnvironmentVariables settingsVariables = variables.getByType(destination.toType());
     String name = EnvironmentVariables.getToolVersionVariable(this.tool);
 
-    VersionIdentifier resolvedVersion = toolRepository.resolveVersion(this.tool, edition, version, this);
-    if (version.isPattern()) {
-      this.context.debug("Resolved version {} to {} for tool {}/{}", version, resolvedVersion, this.tool, edition);
-    }
-    settingsVariables.set(name, resolvedVersion.toString(), false);
+    toolRepository.resolveVersion(this.tool, edition, version, this); // verify that the version actually exists
+    settingsVariables.set(name, version.toString(), false);
     settingsVariables.save();
-    this.context.info("{}={} has been set in {}", name, version, settingsVariables.getSource());
     EnvironmentVariables declaringVariables = variables.findVariable(name);
     if ((declaringVariables != null) && (declaringVariables != settingsVariables)) {
       this.context.warning("The variable {} is overridden in {}. Please remove the overridden declaration in order to make the change affect.", name,

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/npm/NpmBasedCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/npm/NpmBasedCommandlet.java
@@ -1,5 +1,6 @@
 package com.devonfw.tools.ide.tool.npm;
 
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
 import java.util.Set;
@@ -42,6 +43,12 @@ public abstract class NpmBasedCommandlet extends LocalToolCommandlet {
     return true;
   }
 
+  @Override
+  protected boolean isIgnoreMissingSoftwareVersionFile() {
+
+    return true;
+  }
+
   /**
    * @return the package of this tool from the NPM registry.
    */
@@ -58,6 +65,10 @@ public abstract class NpmBasedCommandlet extends LocalToolCommandlet {
   }
 
   protected VersionIdentifier runNpmGetInstalledPackageVersion(String npmPackage) {
+    if (!Files.isDirectory(this.context.getSoftwarePath().resolve("node"))) {
+      this.context.trace("Since node is not installed, also package {} for tool {} cannot be installed.", npmPackage, this.tool);
+      return null;
+    }
     ProcessResult result = runNpm(ProcessMode.DEFAULT_CAPTURE, ProcessErrorHandling.NONE, "list", "-g", npmPackage, "--depth=0");
     if (result.isSuccessful()) {
       List<String> versions = result.getOut();

--- a/cli/src/test/java/com/devonfw/tools/ide/commandlet/InstallCommandletTest.java
+++ b/cli/src/test/java/com/devonfw/tools/ide/commandlet/InstallCommandletTest.java
@@ -4,6 +4,8 @@ import org.junit.jupiter.api.Test;
 
 import com.devonfw.tools.ide.context.AbstractIdeContextTest;
 import com.devonfw.tools.ide.context.IdeContext;
+import com.devonfw.tools.ide.context.IdeTestContext;
+import com.devonfw.tools.ide.tool.repository.DefaultToolRepository;
 import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo;
 import com.github.tomakehurst.wiremock.junit5.WireMockTest;
 
@@ -40,16 +42,42 @@ public class InstallCommandletTest extends AbstractIdeContextTest {
   public void testInstallCommandletRunWithVersionAndVersionIdentifier(WireMockRuntimeInfo wmRuntimeInfo) {
 
     // arrange
-    IdeContext context = newContext(PROJECT_BASIC, wmRuntimeInfo);
+    IdeTestContext context = newContext(PROJECT_BASIC, wmRuntimeInfo);
     InstallCommandlet install = context.getCommandletManager().getCommandlet(InstallCommandlet.class);
     install.tool.setValueAsString("java", context);
-    install.version.setValueAsString("17.0.6", context);
-
+    String version17 = "17.0.6";
+    install.version.setValueAsString(version17, context);
     // act
     install.run();
     // assert
     assertTestInstall(context);
-    assertThat(context.getSoftwarePath().resolve("java/.ide.software.version")).exists().hasContent("17.0.6");
+    assertThat(context).logAtSuccess().hasMessage("Successfully installed java in version " + version17);
+    assertThat(context.getSoftwarePath().resolve("java/.ide.software.version")).exists().hasContent(version17);
+
+    // now we install a different version
+
+    // arrange
+    String version21 = "21.0.8_9";
+    install.version.setValueAsString(version21, context);
+    // act
+    install.run();
+    // assert
+    assertTestInstall(context);
+    assertThat(context).logAtSuccess().hasMessage("Successfully installed java in version " + version21 + " replacing previous version " + version17);
+    assertThat(context.getSoftwarePath().resolve("java/.ide.software.version")).exists().hasContent(version21);
+
+    // now we install the initial version again that should just change the symlink
+
+    // arrange
+    install.version.setValueAsString(version17, context);
+    // act
+    install.run();
+    // assert
+    assertTestInstall(context);
+    assertThat(context).logAtDebug().hasMessage("Version " + version17 + " of tool java is already installed at " + context.getSoftwareRepositoryPath().resolve(
+        DefaultToolRepository.ID_DEFAULT).resolve("java").resolve("java").resolve(version17));
+    assertThat(context).logAtSuccess().hasMessage("Successfully installed java in version " + version17 + " replacing previous version " + version21);
+    assertThat(context.getSoftwarePath().resolve("java/.ide.software.version")).exists().hasContent(version17);
   }
 
   private void assertTestInstall(IdeContext context) {

--- a/cli/src/test/java/com/devonfw/tools/ide/tool/npm/NpmTest.java
+++ b/cli/src/test/java/com/devonfw/tools/ide/tool/npm/NpmTest.java
@@ -6,7 +6,6 @@ import org.junit.jupiter.params.provider.ValueSource;
 import com.devonfw.tools.ide.context.AbstractIdeContextTest;
 import com.devonfw.tools.ide.context.IdeTestContext;
 import com.devonfw.tools.ide.os.SystemInfo;
-import com.devonfw.tools.ide.os.SystemInfoImpl;
 import com.devonfw.tools.ide.os.SystemInfoMock;
 
 /**
@@ -58,7 +57,7 @@ public class NpmTest extends AbstractIdeContextTest {
     commandlet.run();
 
     // assert
-    assertThat(context).logAtInfo().hasMessageContaining("npm" + getBinaryType(context) + " " + getOs(context) + " --version");
+    assertThat(context).logAtInfo().hasMessageContaining("npm " + getOs(context) + " --version");
   }
 
   private void checkInstallation(IdeTestContext context) {
@@ -76,13 +75,5 @@ public class NpmTest extends AbstractIdeContextTest {
       return "mac";
     }
     return "";
-  }
-
-  private String getBinaryType(IdeTestContext context) {
-    String os = "";
-    if (SystemInfoImpl.INSTANCE.isWindows()) {
-      os = "cmd";
-    }
-    return os;
   }
 }

--- a/cli/src/test/resources/ide-projects/basic/_ide/urls/java/java/21.0.8_9/linux_x64.urls
+++ b/cli/src/test/resources/ide-projects/basic/_ide/urls/java/java/21.0.8_9/linux_x64.urls
@@ -1,0 +1,1 @@
+${testbaseurl}/download/java/java/21.0.8_9/java-21.0.8_9.tgz

--- a/cli/src/test/resources/ide-projects/basic/_ide/urls/java/java/21.0.8_9/mac_x64.urls
+++ b/cli/src/test/resources/ide-projects/basic/_ide/urls/java/java/21.0.8_9/mac_x64.urls
@@ -1,0 +1,1 @@
+${testbaseurl}/download/java/java/21.0.8_9/java-21.0.8_9.dmg

--- a/cli/src/test/resources/ide-projects/basic/_ide/urls/java/java/21.0.8_9/windows_x64.urls
+++ b/cli/src/test/resources/ide-projects/basic/_ide/urls/java/java/21.0.8_9/windows_x64.urls
@@ -1,0 +1,1 @@
+${testbaseurl}/download/java/java/21.0.8_9/java-21.0.8_9.zip

